File: library/src/main/java/com/shxhzhxx/imageloader/BitmapLoader.java
package com.shxhzhxx.imageloader;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.BitmapRegionDecoder;
import android.graphics.Rect;
import android.os.Build;
import android.support.annotation.MainThread;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;
import android.text.TextUtils;
import android.util.Log;
import android.util.LruCache;

import com.shxhzhxx.urlloader.MultiObserverTaskManager;
import com.shxhzhxx.urlloader.UrlLoader;

import java.io.File;
import java.io.IOException;

public class BitmapLoader extends MultiObserverTaskManager<BitmapLoader.ProgressObserver> {
    private static final String TAG = "BitmapLoader";
    private static final int MEM_SCALE = 1024; //mem unit: kilo byte

    private static final int CONFIG_MASK = 0x00FF;

    /**
     * By default,bitmap will be resized so that its relatively longer edges can match the desired size.
     * With this configuration, the bitmap will be resized so that its relatively shorter edges can match the desired size and longer edges will be cropped.
     * Currently only the JPEG and PNG formats are supported
     */
    public static final int CONFIG_CENTER_CROP = 1;


    public abstract static class ProgressObserver {
        /**
         * @param bitmap immutable bitmap
         */
        public void onComplete(Bitmap bitmap) {
        }

        public void onFailed() {
        }

        public void onCanceled() {
        }
    }


    @Nullable
    public static Bitmap loadBitmap(File file, int width, int height) {
        if (height <= 0 && width <= 0) {
            Log.e(TAG, "load bitmap without compress :" + file.getAbsolutePath());
        }
        if (height <= 0) {
            height = Integer.MAX_VALUE;
        }
        if (width <= 0) {
            width = Integer.MAX_VALUE;
        }
        BitmapFactory.Options opt = new BitmapFactory.Options();
        opt.inJustDecodeBounds = true;
        BitmapFactory.decodeFile(file.getAbsolutePath(), opt);

        double inSampleSize = Math.max((double) opt.outWidth / (double) width, (double) opt.outHeight / (double) height);
        opt = new BitmapFactory.Options();
        opt.inSampleSize = (int) inSampleSize;
        opt.inJustDecodeBounds = false;
        Bitmap bitmap = BitmapFactory.decodeFile(file.getAbsolutePath(), opt); //初步压缩分辨率
        if (bitmap == null)
            return null;
        if (inSampleSize <= 1) {
            return bitmap;
        }
        inSampleSize = Math.max((double) bitmap.getWidth() / (double) width, (double) bitmap.getHeight() / (double) height);
        return Bitmap.createScaledBitmap(bitmap, (int) (bitmap.getWidth() / inSampleSize), (int) (bitmap.getHeight() / inSampleSize), true); //进一步压缩分辨率
    }

    @Nullable
    public static Bitmap loadBitmapCrop(File file, int width, int height) {
        if (width <= 0 || height <= 0) {
            Log.e(TAG, "loadBitmapCrop: crop need explicit width and height");
            return loadBitmap(file, width, height);
        }
        BitmapRegionDecoder decoder;
        try {
            decoder = BitmapRegionDecoder.newInstance(file.getAbsolutePath(), false);
        } catch (IOException e) {
            Log.e(TAG, "loadBitmapCrop IOException:" + e.getMessage());
            return null;
        }
        int rawW = decoder.getWidth();
        int rawH = decoder.getHeight();
        double inSampleSize = Math.min((double) rawW / (double) width, (double) rawH / (double) height);
        BitmapFactory.Options opt = new BitmapFactory.Options();
        opt.inSampleSize = (int) inSampleSize;
        int left = (int) (rawW / 2 - width * inSampleSize / 2);
        int right = (int) (rawW / 2 + width * inSampleSize / 2);
        int top = (int) (rawH / 2 - height * inSampleSize / 2);
        int bottom = (int) (rawH / 2 + height * inSampleSize / 2);
        Bitmap bitmap = decoder.decodeRegion(new Rect(left, top, right, bottom), opt);
        if (bitmap == null)
            return null;
        if (inSampleSize <= 1) {
            return bitmap;
        }
        inSampleSize = Math.min((double) bitmap.getWidth() / (double) width, (double) bitmap.getHeight() / (double) height);
        return Bitmap.createScaledBitmap(bitmap, (int) (bitmap.getWidth() / inSampleSize), (int) (bitmap.getHeight() / inSampleSize), true);
    }

    private UrlLoader mUrlLoader;
    private LruCache<String, Bitmap> mMemoryCache;

    public BitmapLoader(@NonNull File cachePath) {
        super(5);
        mUrlLoader = new UrlLoader(cachePath, 50 * 1024 * 1024, 5);
        mMemoryCache = new LruCache<String, Bitmap>((int) (Runtime.getRuntime().maxMemory() / (MEM_SCALE * 8))) {
            @Override
            protected int sizeOf(String key, Bitmap value) {
                return value.getByteCount() / MEM_SCALE;
            }

            @Override
            protected void entryRemoved(boolean evicted, String key, Bitmap oldValue, Bitmap newValue) {
                /* The old bitmap may now be used by other views, so the oldValue can not be reclaimed here.
                 *  actually, calling bitmap.recycle() is needless on Android 3.0 and higher
                 */
            }
        };
    }

    @MainThread
    private String getKey(String path, int width, int height, int config) {
        return mUrlLoader.md5(mUrlLoader.md5(mUrlLoader.md5(mUrlLoader.md5(path) + width) + height) + (config & CONFIG_MASK));
    }

    public UrlLoader getUrlLoader() {
        return mUrlLoader;
    }

    public int load(final String path, @Nullable String tag, final int width, final int height, final int config, ProgressObserver observer) {
        if (TextUtils.isEmpty(path) || width < 0 || height < 0) {
            Log.e(TAG, TAG + ".load: invalid params");
            if (observer != null)
                observer.onFailed();
            return -1;
        }
        File file = new File(path);
        if (file.exists()) {
            return load(file, tag, width, height, config, observer);
        }
        String key = getKey(path, width, height, config);
        Bitmap bitmap = mMemoryCache.get(key);
        if (bitmap != null) {
            if (observer != null)
                observer.onComplete(bitmap);
            return -2;
        }
        int id = start(key, tag, observer, new TaskBuilder() {
            @Override
            public Task build() {
                return new WorkThread(path, width, height, config);
            }
        });
        if (id < 0 && observer != null) {
            observer.onFailed();
        }
        return id;
    }

    public int load(String path, int width, int height, int config, ProgressObserver observer) {
        return load(path, null, width, height, config, observer);
    }

    public int load(final File file, @Nullable String tag, final int width, final int height, final int config, ProgressObserver observer) {
        if (file == null || !file.exists() || width < 0 || height < 0) {
            Log.e(TAG, TAG + ".load: invalid params");
            if (observer != null)
                observer.onFailed();
            return -1;
        }
        String key = getKey(file.getAbsolutePath(), width, height, config);
        Bitmap bitmap = mMemoryCache.get(key);
        if (bitmap != null) {
            if (observer != null)
                observer.onComplete(bitmap);
            return -2;
        }
        int id = start(key, tag, observer, new TaskBuilder() {
            @Override
            public Task build() {
                return new WorkThread(file, width, height, config);
            }
        });
        if (id < 0 && observer != null) {
            observer.onFailed();
        }
        return id;
    }

    public int load(final File file, final int width, final int height, final int config, ProgressObserver observer) {
        return load(file, null, width, height, config, observer);
    }

    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR1)
    public void clearMemCache() {
        mMemoryCache.trimToSize(-1);
    }

    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR1)
    public void trimMemory(boolean resizeCache) {
        Log.d(TAG, "trimMemory  resizeCache:" + resizeCache);
        int size = mMemoryCache.size();
        if (size < 2) // can't be smaller than 1
            return;
        if (resizeCache && Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            mMemoryCache.resize(size / 2);
        } else {
            mMemoryCache.trimToSize(size / 2);
        }
    }

    private class WorkThread extends Task {
        private volatile File mFile;
        private Bitmap mBitmap;
        private int mHeight, mWidth, mConfig;
        private volatile int mDownloadId;

        @MainThread
        WorkThread(String url, int width, int height, int config) {
            super(BitmapLoader.this.getKey(url, width, height, config));
            mWidth = width;
            mHeight = height;
            mConfig = config;

            //work thread doesn't start yet ,no need atomic operation
            //noinspection NonAtomicOperationOnVolatileField
            mDownloadId = mUrlLoader.load(url, new UrlLoader.ProgressObserver() {
                @Override
                public void onComplete(File file) {
                    synchronized (WorkThread.this) {
                        mFile = file;
                        mDownloadId = -1;
                        WorkThread.this.notify();
                    }
                }

                @Override
                public void onCanceled() {
                    synchronized (WorkThread.this) {
                        if (mDownloadId >= 0) {
                            WorkThread.this.notify();
                            mDownloadId = -1;
                        }
                    }
                }

                @Override
                public void onFailed() {
                    synchronized (WorkThread.this) {
                        mDownloadId = -1;
                        WorkThread.this.notify();
                    }
                }
            });
        }

        WorkThread(File file, int width, int height, int config) {
            super(BitmapLoader.this.getKey(file.getAbsolutePath(), width, height, config));
            mFile = file;
            mWidth = width;
            mHeight = height;
            mConfig = config;
            mDownloadId = -1;
        }

        /**
         * check {@link #mDownloadId} in main thread, don't need to synchronize.
         * {@link #wait()} will be aroused by interrupt()
         */
        @Override
        @MainThread
        protected void onCanceled() {
            if (mDownloadId >= 0 && mUrlLoader.cancel(mDownloadId)) {
                mDownloadId = -1;
            }
            for (ProgressObserver observer : getObservers()) {
                if (observer != null)
                    observer.onCanceled();
            }
        }

        @Override
        protected void onObserverUnregistered(ProgressObserver observer) {
            if (observer != null)
                observer.onCanceled();
        }

        @Nullable
        private Bitmap load() {
            try {
                if ((mConfig & CONFIG_CENTER_CROP) != 0) {
                    return loadBitmapCrop(mFile, mWidth, mHeight);
                } else {
                    return loadBitmap(mFile, mWidth, mHeight);
                }
            } catch (Exception e) {
                Log.e(TAG, "load Exception: " + e.getMessage());
                return null;
            }
        }

        @Override
        protected void doInBackground() {
            synchronized (WorkThread.this) {
                while (!isCanceled() && mDownloadId >= 0) {
                    try {
                        WorkThread.this.wait();
                    } catch (InterruptedException ignore) {
                    }
                }
            }
            if (isCanceled())
                return;
            if (mFile != null && mFile.exists() && (mBitmap = load()) != null) {
                if ((mBitmap.getByteCount() / MEM_SCALE) < (mMemoryCache.maxSize() / 2))
                    mMemoryCache.put(getKey(), mBitmap);
                setPostResult(new Runnable() {
                    @Override
                    public void run() {
                        for (ProgressObserver observer : getObservers()) {
                            if (observer != null)
                                observer.onComplete(mBitmap);
                        }
                    }
                });
            } else {
                setPostResult(new Runnable() {
                    @Override
                    public void run() {
                        for (ProgressObserver observer : getObservers()) {
                            if (observer != null)
                                observer.onFailed();
                        }
                    }
                });
            }
        }
    }
}

File: library/src/main/java/com/shxhzhxx/imageloader/BitmapLoader.kt
package com.shxhzhxx.imageloader

import android.content.ContentResolver
import android.graphics.*
import android.net.Uri
import android.os.Build
import android.util.Log
import android.util.LruCache
import androidx.annotation.IntRange
import androidx.annotation.RequiresApi
import androidx.exifinterface.media.ExifInterface
import com.shxhzhxx.urlloader.TaskManager
import com.shxhzhxx.urlloader.UrlLoader
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStream
import java.util.*
import kotlin.math.max
import kotlin.math.min

const val ROUND_CIRCLE = -1
private const val TAG = "BitmapLoader"
private const val MEM_SCALE = 1024


class BitmapLoader(private val contentResolver: ContentResolver, private val fileCachePath: File) : TaskManager<BitmapLoader.Holder, Bitmap>() {
    class Holder(
            val onLoad: ((Bitmap) -> Unit)? = null,
            val onFailure: (() -> Unit)? = null,
            val onCancel: (() -> Unit)? = null
    )

    val urlLoader = UrlLoader(fileCachePath, 50 * 1024 * 1024)
    private val memoryCache = object : LruCache<Params, Bitmap>((max(1, Runtime.getRuntime().maxMemory() / MEM_SCALE / 8)).toInt()) {
        override fun sizeOf(key: Params, value: Bitmap) = value.byteCount / MEM_SCALE
    }

    /**
     * @param path could be url, [File.getAbsolutePath] , [Uri.toString]
     * */
    @JvmOverloads
    fun load(path: String, @IntRange(from = 0) width: Int = 0, @IntRange(from = 0) height: Int = 0, centerCrop: Boolean = true, roundingRadius: Int = 0, precisely: Boolean = false, tag: Any? = null,
             onLoad: ((Bitmap) -> Unit)? = null,
             onFailure: (() -> Unit)? = null,
             onCancel: (() -> Unit)? = null): Int {
        val params = Params(path, width, height, centerCrop, roundingRadius, precisely)
        return asyncStart(params, { Worker(params) }, tag, Holder(onLoad, onFailure, onCancel)).also { id ->
            if (id < 0) {
                onFailure?.invoke()
            }
        }
    }

    fun syncLoad(path: String, canceled: () -> Boolean, @IntRange(from = 0) width: Int = 0, @IntRange(from = 0) height: Int = 0, centerCrop: Boolean = true, roundingRadius: Int = 0, precisely: Boolean = false): Bitmap? {
        val params = Params(path, width, height, centerCrop, roundingRadius, precisely)
        return syncStart(params, { Worker(params) }, canceled)
    }

    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    fun resizeMemCache(@IntRange(from = 1) maxSize: Int) {
        memoryCache.resize(maxSize)
    }

    fun clearMemCache() {
        memoryCache.evictAll()
    }

    fun trimMemory(resizeCache: Boolean) {
        val size = memoryCache.size()
        when {
            size < 2 -> return
            resizeCache && Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP -> resizeMemCache(size / 2)
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1 -> memoryCache.trimToSize(size / 2)
            else -> clearMemCache()
        }
    }

    private inner class Worker(private val params: Params) : Task(params) {
        override fun doInBackground(): Bitmap? {
            if (isCancelled)
                return null
            val bitmap = memoryCache[params] ?: try {
                fun inputStream() = (contentResolver.openInputStream(Uri.parse(params.path))
                        ?: throw FileNotFoundException("Unable to create stream"))
                decodeBitmap({ inputStream() }, params, inputStream().readRotateThenClose())
            } catch (e: FileNotFoundException) {
                val f = File(params.path)
                (if (f.exists()) f else urlLoader.syncLoad(params.path, { isCancelled || (allSyncCanceled && asyncObservers.isEmpty()) }))?.decodeBitmap(params)
            }?.also {
                if (it.byteCount / MEM_SCALE < memoryCache.size() / 2) {
                    memoryCache.put(params, it)
                }
            }

            postResult = if (bitmap != null) Runnable {
                observers.forEach { it?.onLoad?.invoke(bitmap) }
            } else Runnable {
                observers.forEach { it?.onFailure?.invoke() }
            }
            return bitmap
        }

        override fun onObserverUnregistered(observer: Holder?) {
            observer?.onCancel?.invoke()
        }

        override fun onCancel() {
            observers.forEach { it?.onCancel?.invoke() }
        }
    }


    private fun InputStream.readRotateThenClose() = use {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            return@use try {
                ExifInterface(this).readRotate()
            } catch (e: IOException) {
                0
            }
        } else {
            val file = File(fileCachePath, urlLoader.md5(UUID.randomUUID().toString()))
            return@use try {
                file.outputStream().use { copyTo(it) }
                file.readRotate()
            } catch (e: IOException) {
                0
            } finally {
                file.delete()
            }
        }
    }

    private fun decodeBitmap(getInputStream: () -> InputStream, params: Params, rotate: Int): Bitmap? {
        fun decodeStream(inputStream: InputStream, opts: BitmapFactory.Options) =
                inputStream.use {
                    try {
                        BitmapFactory.decodeStream(it, null, opts)
                    } catch (e: Throwable) {
                        null
                    }
                }

        val centerCrop = params.centerCrop && params.height > 0 && params.width > 0
        if (params.height <= 0 && params.width <= 0) {
            Log.e(TAG, "load bitmap without compress")
        }
        val opts = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        decodeStream(getInputStream(), opts)
        val (dstWidth, dstHeight) = when (rotate) {
            90, 270 -> params.height to params.width
            else -> params.width to params.height
        }
        val (out, dst) = listOf(opts.outHeight to dstHeight, opts.outWidth to dstWidth).filter { it.second > 0 }
                .run { return@run if (centerCrop) minBy { it.first.toFloat() / it.second } else maxBy { it.first.toFloat() / it.second } }
                ?: 0 to 0
        opts.inJustDecodeBounds = false
        opts.inSampleSize = if (dst == 0) 0 else out / dst

        return (if (!centerCrop) {
            decodeStream(getInputStream(), opts)?.let { bitmap ->
                val resizeFactor = listOf(dstWidth to bitmap.width, dstHeight to bitmap.height).filter { it.first > 0 && it.second > 0 }
                        .map { it.first.toFloat() / it.second }.min() ?: return@let bitmap
                return@let if (!params.precisely || resizeFactor == 1f) bitmap else {
                    Bitmap.createScaledBitmap(bitmap, (bitmap.width * resizeFactor).toInt(), (bitmap.height * resizeFactor).toInt(), true)
                }
            }
        } else {
            getInputStream().use { inputStream ->
                try {
                    val inSampleSize = out.toFloat() / dst
                    val bitmap = BitmapRegionDecoder.newInstance(inputStream, false).decodeRegion(Rect(
                            ((opts.outWidth / 2 - dstWidth * inSampleSize / 2).toInt()),
                            ((opts.outHeight / 2 - dstHeight * inSampleSize / 2).toInt()),
                            ((opts.outWidth / 2 + dstWidth * inSampleSize / 2).toInt()),
                            ((opts.outHeight / 2 + dstHeight * inSampleSize / 2).toInt())), opts)
                            ?: return null

                    val resizeFactor = listOf(dstWidth to bitmap.width, dstHeight to bitmap.height).find { it.first > 0 && it.second > 0 }
                            ?.let { it.first.toFloat() / it.second } ?: return@use bitmap
                    if (!params.precisely || resizeFactor == 1f) bitmap else Bitmap.createScaledBitmap(bitmap, dstWidth, dstHeight, true)
                } catch (e: IOException) {
                    null
                }
            }
        })?.let { bitmap ->
            if (rotate == 0) bitmap else Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, Matrix().apply { postRotate(rotate.toFloat()) }, true)
        }?.let { bitmap ->
            when {
                params.roundingRadius == 0 -> bitmap
                params.roundingRadius < 0 -> bitmap.toRounded(min(bitmap.width, bitmap.height).toFloat() / 2)
                else -> bitmap.toRounded(params.roundingRadius.toFloat())
            }
        }
    }

    private fun Bitmap.toRounded(radius: Float) = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888).also { bitmap ->
        Canvas(bitmap).also { canvas ->
            canvas.drawRoundRect(RectF(0f, 0f, bitmap.width.toFloat(), bitmap.height.toFloat()), radius, radius, Paint().also { paint ->
                paint.isAntiAlias = true
                paint.shader = BitmapShader(this, Shader.TileMode.CLAMP,
                        Shader.TileMode.CLAMP)
            })
            canvas.setBitmap(null)
        }
    }

    private fun ExifInterface.readRotate() = when (getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL)) {
        ExifInterface.ORIENTATION_ROTATE_90 -> 90
        ExifInterface.ORIENTATION_ROTATE_180 -> 180
        ExifInterface.ORIENTATION_ROTATE_270 -> 270
        else -> 0
    }

    private fun File.readRotate() = try {
        ExifInterface(absolutePath).readRotate()
    } catch (e: IOException) {
        0
    }


    private fun File.decodeBitmap(params: Params) = try {
        decodeBitmap({ inputStream() }, params, readRotate())
    } catch (e: FileNotFoundException) {
        null
    }
}

data class Params(val path: String, val width: Int, val height: Int, val centerCrop: Boolean, val roundingRadius: Int, val precisely: Boolean)

